test-data/unit/check-selftype.test

```diff
@@ -0,0 +1,219 @@
+
+[case testSelfTypeStaticAccess]
+from typing import TypeVar
+
+T = TypeVar('T')
+class A:
+    def copy(self: T) -> T: ...
+
+class B(A):
+    pass
+
+# Partial types appear on reveal_type; unrelated to self type
+def f(a: A) -> None: ...
+f(A.copy(A()))
+f(A.copy(B()))
+f(B.copy(B()))
+
+# TODO: make it an error
+# f(B.copy(A()))
+
+def g(a: B) -> None: ...
+g(A.copy(A()))  # E: Argument 1 to "g" has incompatible type "A"; expected "B"
+g(A.copy(B()))
+g(B.copy(B()))
+
+[case testSelfTypeReturn]
+# flags: --hide-error-context
+from typing import TypeVar, Type
+
+T = TypeVar('T')
+
+def _type(self: T) -> Type[T]: ...
+
+class A:
+    def copy(self: T) -> T:
+        if B():
+            return A()  # E: Incompatible return value type (got "A", expected "T")
+        elif A():
+            return B()  # E: Incompatible return value type (got "B", expected "T")
+        return _type(self)()
+
+class B(A):
+    pass
+
+class C:
+    def __init__(self, a: int) -> None: pass
+
+    def copy(self: T) -> T:
+        if self:
+            return _type(self)(1)
+        else:
+            return _type(self)()  # E: Too few arguments for "C"
+
+
+
+[case testSelfTypeClass]
+# flags: --hide-error-context
+from typing import TypeVar, Type
+
+T = TypeVar('T')
+
+class A:
+    @classmethod
+    def new(cls: Type[T]) -> T:
+        return cls()
+
+class B(A):
+    pass
+
+class C:
+    def __init__(self, a: int) -> None: pass
+
+    @classmethod
+    def new(cls: Type[T]) -> T:
+        if cls:
+            return cls(1)
+        else:
+            return cls()  # E: Too few arguments for "C"
+
+
+reveal_type(A.new)  # E: Revealed type is 'def () -> __main__.A'
+reveal_type(B.new)  # E: Revealed type is 'def () -> __main__.B'
+reveal_type(A.new())  # E: Revealed type is '__main__.A'
+reveal_type(B.new())  # E: Revealed type is '__main__.B'
+
+[builtins fixtures/classmethod.pyi]
+
+[case testSelfTypeOverride]
+from typing import TypeVar, cast
+ 
+T = TypeVar('T')
+ 
+class A:
+    def copy(self: T) -> T: ...
+
+class B(A):
+    pass
+
+class C(A):
+    def copy(self: T) -> T: ...
+
+reveal_type(C().copy)  # E: Revealed type is 'def () -> __main__.C'
+reveal_type(C().copy())  # E: Revealed type is '__main__.C'
+reveal_type(cast(A, C()).copy)  # E: Revealed type is 'def () -> __main__.A'
+reveal_type(cast(A, C()).copy())  # E: Revealed type is '__main__.A' 
+
+[case testSelfTypeSuper]
+# flags: --hide-error-context
+from typing import TypeVar, cast
+ 
+T = TypeVar('T')
+ 
+class A:
+    def copy(self: T) -> T: ...
+
+class B(A):
+    def copy(self: T) -> T:
+        reveal_type(self)  # E: Revealed type is 'T`-1'
+        reveal_type(super().copy)  # E: Revealed type is 'def () -> T`-1'
+        return super().copy()
+
+[case testSelfTypeRecursiveBinding]
+# flags: --hide-error-context
+from typing import TypeVar, Callable, Type
+ 
+T = TypeVar('T')
+class A:
+    def copy(self: T, factory: Callable[[T], T]) -> T:
```

Comment:
Overriding this method should be forbidden with a type variable that has a subclass bound. Example:

```
T = TypeVar('T')

class A:
    def copy(self: T, factory: Callable[[T], T]) -> T:
        return f(self)

S = TypeVar('S', bound='B')

class B(A):
    x = 1

    def copy(self: S, factory: Callable[[S], S]) -> S:   # Not good
        factory(self).x
        return self

a = B()  # type: A
def f(a: A) -> A: return A()
a.copy(f)  # Error
```

Can you add this test case for such an unsafe override (or a TODO if it doesn't work yet)?

If the bound remains the same, this would likely be okay. Also, if `T` is not used in a method argument type (only used in the return value type), the bound can vary covariantly. This would be another good thing to test.